django/db/migrations/loader.py

```diff
@@ -249,13 +278,63 @@ def build_graph(self):
         for key, migration in self.disk_migrations.items():
             self.add_external_dependencies(key, migration)
         # Carry out replacements where possible and if enabled.
+
+        @memoize_and_check_cycles
+        def has_been_applied(target):
+            # we have two ways that a target can be applied
+            # 1. we have it in self.applied_migrations (registered in the DB)
+            # 2. it is a squashed migration, and its replaces have been applied
+            if target in self.applied_migrations:
+                return True
+            if target in self.replacements:
+                # this has a replaces key, let's look up whether those have been applied
+                migration = self.replacements[target]
+                return all(has_been_applied(target) for target in migration.replaces)
+            return False
+
         if self.replace_migrations:
-            for key, migration in self.replacements.items():
-                # Get applied status of each of this migration's replacement
-                # targets.
-                applied_statuses = [
-                    (target in self.applied_migrations) for target in migration.replaces
-                ]
+            # this tracks nodes that were replaced (this is to handle squashing of squashing)
+
+            # migration_work[key] = [migration, visited, finished_replacement]
+            # this is mainly used to track cyclical dependencies and which migrations we have
+            # finished replacing
+            migration_work = {
+                key: [migration, False, False]
+                for key, migration in self.replacements.items()
+            }
+
+            def do_replacement(key):
+                # Toggle here to test between old code path and new one
+                NEW_WAY = True
```

Comment:
a helper boolean for me to toggle between the old and new ways when exploring how to write a regression test (turns out the existing test would fail sporadically in practice)